### DISCUSSION TITLE
RP2 Pico SPI breaking change - a warning

### DISCUSSION THREAD

**Discussion by peterhinch** [micropython/micropython#9428] [General]
Formerly (firmware up to August 8th) I issued:
```python
spi = SPI(0, baudrate=30_000_000)
```
which worked on GPIO 4, 6 and 7. With today's daily build it is now necessary to specify these pins explicitly:
```python
spi = SPI(0, sck=Pin(6), mosi=Pin(7), miso=Pin(4), baudrate=30_000_000)
```

**Comment by robert-hh** [comment]
It was PR #9224 which caused the change. Now different people complain. Lesson learned: always specify the pins. That’s also good for documentation.

**Comment by peterhinch** [comment]
I wrote my code when RP2 was new. The docs I had at the time didn't show alternative pinouts for SPI(0) so I used the pins shown on the diagram and it worked.

While I agree with your comment it seems an odd decision to change the defaults.

**Reply by jimmo** [reply]
Does this mean that the official pinout diagrams (and therefore the meaning of the "default pins") changed between the release and now?

I was under the impression that the existing defaults were just "wrong"... i.e. they didn't match any official guidelines and were effectively arbitrary and what we're doing now is using the defaults as defined by the Pico SDK. But nothing is ever simple!

**Reply by dpgeorge** [reply]
Yes, from what I understand the default pins now use what the pico-sdk defines as default pins. 

It's unfortunate that it changed the defaults from what we had at the beginning...

**Reply by dpgeorge** [reply]
Looking through some old files on my PC, I found a Pico pinout that does not have defaults for SPI or I2C (but does have default for UART).  So what may have happened is that I used this pinout for the original Pico board definition, and since it didn't have defaults for SPI or I2C I just used the lowest numbered GPIOs.

**Comment by dhalbert** [comment]
In CircuitPython, we started to assign default pins for the RPi Pico ,and then gave up, because there was more than one "standard", and there were no board labels. We do provide default pins when the silkscreen identifies them, or there is some extremely strong standard (e.g., Arduino-shaped-board SPI pins).

Discussions here:
https://github.com/adafruit/circuitpython/issues/4121
https://github.com/adafruit/circuitpython/issues/5041

**Reply by jimmo** [reply]
"We decided that trying to ameliorate board designs that fail to label default pins is beyond our purview."

Well I definitely agree there. I was hoping that deferring to the Pico SDK would be the simple solution but yeah now I see. Thanks for sharing @dhalbert 

(There's a broader topic here and a long-standing TODO about the concept of "board" in general for the non-stm32 ports on MicroPython... i.e. the equivalent of Pyboards "X" and "Y" busses and labelling pins etc, much like CircuitPython's `board` module)

**Comment by peterhinch** [comment]
FWIW I used the pinout diagram still shown on [this page](https://shop.mchobby.be/fr/pico-rp2040/2434-pico-w-wireless-rp2040-2-coeurs-wifi-bluetooth-3232100024342.html).

**Reply by DeeJay** [reply]
> FWIW I used the pinout diagram still shown on [this page](https://shop.mchobby.be/fr/pico-rp2040/2434-pico-w-wireless-rp2040-2-coeurs-wifi-bluetooth-3232100024342.html).


There is nothing necessarily wrong with that seller-provided information, but it might be useful to refer to the 'canonical' datasheets (including similar pinout diagrams) published by Raspberry Pi themselves:
[For RPi Pico - ](https://datasheets.raspberrypi.com/pico/pico-datasheet.pdf)
[For RPi PicoW - ](https://datasheets.raspberrypi.com/picow/pico-w-datasheet.pdf)

Both diagrams also appear in the official documentation overview - [https://www.raspberrypi.com/documentation/microcontrollers/raspberry-pi-pico.html](https://www.raspberrypi.com/documentation/microcontrollers/raspberry-pi-pico.html)
